File: sudoq-app/sudoqmodel/src/main/java/de/sudoq/model/solverGenerator/solution/LeftoverNoteDerivation.java
package de.sudoq.model.solverGenerator.solution;

import java.util.ArrayList;
import java.util.List;

import de.sudoq.model.actionTree.Action;
import de.sudoq.model.actionTree.NoteActionFactory;
import de.sudoq.model.solvingAssistant.HintTypes;
import de.sudoq.model.sudoku.Constraint;
import de.sudoq.model.sudoku.Field;
import de.sudoq.model.sudoku.Position;
import de.sudoq.model.sudoku.Sudoku;
import de.sudoq.model.sudoku.Utils;
import de.sudoq.model.sudoku.Utils.ConstraintShape;

/**
 * Created by timo on 04.10.16.
 */
public class LeftoverNoteDerivation extends SolveDerivation {


    private Constraint constraint;
    private int note;

    private List<Action> actionlist;


    public LeftoverNoteDerivation(Constraint c, int note) {
        super(HintTypes.LeftoverNote);
        constraint = c;
        this.note = note;
        this.actionlist = new ArrayList<>();
        hasActionListCapability = true;
    }

    public Constraint getConstraint(){ return constraint; }
    public int getNote(){ return note; }

    public ConstraintShape getConstraintShape(){
        return Utils.getGroupShape(constraint);
    }


    @Override
    public List<Action> getActionList(Sudoku sudoku){
        NoteActionFactory af = new NoteActionFactory();

        for (Position p : constraint ) {
            Field f = sudoku.getField(p);
            if (f.isNoteSet(note) && f.isNotSolved())
                actionlist.add(af.createAction(note, f));
        }
        return actionlist;
    }
}

File: sudoq-app/sudoqmodel/src/main/kotlin/de/sudoq/model/solverGenerator/solution/LeftoverNoteDerivation.kt
package de.sudoq.model.solverGenerator.solution

import de.sudoq.model.actionTree.Action
import de.sudoq.model.actionTree.NoteActionFactory
import de.sudoq.model.solvingAssistant.HintTypes
import de.sudoq.model.sudoku.Constraint
import de.sudoq.model.sudoku.Sudoku
import de.sudoq.model.sudoku.Utils
import de.sudoq.model.sudoku.getGroupShape

/**
 * Created by timo on 04.10.16.
 */
class LeftoverNoteDerivation(val constraint: Constraint, val note: Int) :
    SolveDerivation(HintTypes.LeftoverNote) {

    private val actionlist: MutableList<Action> = ArrayList()

    val constraintShape: Utils.ConstraintShape
        get() = getGroupShape(constraint)

    override fun getActionList(sudoku: Sudoku): List<Action> {
        val af = NoteActionFactory()
        for (p in constraint) {
            val f = sudoku.getCell(p)
            if (f.isNoteSet(note) && f.isNotSolved) actionlist.add(af.createAction(note, f))
        }
        return actionlist
    }

    init {
        hasActionListCapability = true
    }
}